src/parse/text/simple_extensions/types.rs

```diff
@@ -0,0 +1,1773 @@
+            nullable,
+        }
+    }
+
+    /// Create a new list type
+    pub fn list(element_type: ConcreteType, nullable: bool) -> ConcreteType {
+        ConcreteType {
+            kind: ConcreteTypeKind::List(Box::new(element_type)),
+            nullable,
+        }
+    }
+
+    /// Create a new struct type (ordered fields without names)
+    pub fn r#struct(field_types: Vec<ConcreteType>, nullable: bool) -> ConcreteType {
+        ConcreteType {
+            kind: ConcreteTypeKind::Struct(field_types),
+            nullable,
+        }
+    }
+
+    /// Create a new map type
+    pub fn map(key_type: ConcreteType, value_type: ConcreteType, nullable: bool) -> ConcreteType {
+        ConcreteType {
+            kind: ConcreteTypeKind::Map {
+                key: Box::new(key_type),
+                value: Box::new(value_type),
+            },
+            nullable,
+        }
+    }
+
+    /// Create a new named struct type (nstruct - ordered fields with names)
+    pub fn named_struct(
+        field_names: Vec<String>,
+        field_types: Vec<ConcreteType>,
+        nullable: bool,
+    ) -> ConcreteType {
+        ConcreteType {
+            kind: ConcreteTypeKind::NamedStruct {
+                field_names,
+                field_types,
+            },
+            nullable,
+        }
+    }
+
+    /// Check if this type matches another type exactly
+    pub fn matches(&self, other: &ConcreteType) -> bool {
+        self == other
+    }
+
+    /// Check if this type is compatible with another type (considering nullability)
+    pub fn is_compatible_with(&self, other: &ConcreteType) -> bool {
+        // Types must match exactly, but nullable types can accept non-nullable values
+        self.kind == other.kind && (self.nullable || !other.nullable)
+    }
+}
+
+impl fmt::Display for ConcreteType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.kind)?;
+        if self.nullable {
+            write!(f, "?")?;
+        }
+        Ok(())
+    }
+}
+
+impl From<ConcreteType> for RawType {
+    fn from(val: ConcreteType) -> Self {
+        match val.kind {
+            ConcreteTypeKind::NamedStruct {
+                field_names,
+                field_types,
+            } => {
+                let mut map = serde_json::Map::new();
+                for (name, ty) in field_names.into_iter().zip(field_types.into_iter()) {
+                    if let Some(v) = map.insert(name, serde_json::Value::String(ty.to_string())) {
+                        // This should not happen - you should not have
+                        // duplicate field names in a NamedStruct
+                        panic!("duplicate value '{v:?}' in NamedStruct");
+                    }
+                }
+                RawType::Object(map)
+            }
+            _ => RawType::String(val.to_string()),
+        }
+    }
+}
+
+fn expect_integer_param(
```

Comment:
```suggestion
/// Extract and validate an integer parameter for a built-in type.
///
/// For `DECIMAL<10,2>`, this validates that `10` (index 0) and `2` (index 1) are integers
/// within their required ranges (precision 1-38, scale 0-precision).
///
/// - `type_name`: Type being validated (for error messages, e.g., "DECIMAL")
/// - `index`: Parameter position (0-based, e.g., 0 for precision, 1 for scale)
/// - `param`: The parameter to validate
/// - `range`: Optional bounds to enforce (e.g., `Some(1..=38)` for precision)
fn expect_integer_param(
```

I found this function a bit confusing. It might be the case that some amount of refactoring could make it more clear from its signature alone, but considering that this function is private, it seems appropriate to leave that potential refactor for later.